### DISCUSSION TITLE
Look into dropping jQuery UI

### DISCUSSION THREAD

**Discussion by javierjulio** [activeadmin/activeadmin#7874] [General]
The jQuery UI library is massive and isn't seeing a lot of development. We should look into where its being used (e.g. calendar, tabs, modal) and see about replacing it with our own JS or smaller libraries dedicated for each.

For a date picker: https://github.com/chmln/flatpickr

**Comment by varyonic** [comment]
See also #4034 re datepickers

**Comment by rystraum** [comment]
I would like to help out on this. For now, we need a list on where jQuery is used, yes?

**Comment by deivid-rodriguez** [comment]
I assume [all of the jquery-ui components we currently vendorize](https://github.com/activeadmin/activeadmin/tree/3c7fac35e2f0912d8e4fd583e5e0512de7450d1a/vendor/assets/javascripts/jquery-ui) use jquery under the hood. So we would need to replace each of these components with a standalone solution that does not depend on jquery. It's an ambitious task.

**Comment by rystraum** [comment]
How open are you to replacing them with React-based components? That would require adding webpacker as a dependency. Just want to set the boundaries on what we can replace the vendorized libraries with.

**Comment by deivid-rodriguez** [comment]
I'm pretty opened to replacing our assets with more modern solutions, not sure about others. Perhaps the task could be subdivided in two subtasks:

* Raplace `sprockets` based assets with `webpacker` based assets, while keeping the current implementation.
* Once we have `webpacker` in place, it should be way easier to replace the current components with more modern implementations, I think.

Word of caution though for the first task: the current state of the art around "webpacker + rails engines" is "pretty tricky". At least last time I checked was by no means straightforward how to set it up.

**Comment by vfonic** [comment]
Ping @rystraum! In case you missed it, I just found this PR that you might be interested in having a look/try: #5855.

**Comment by rystraum** [comment]
Thanks for the ping @vfonic. I was sort of stuck with this because I was following this thread https://github.com/rails/webpacker/issues/348. It didn't cross my mind that we can just write all the files to `app/javascript/activeadmin` upon running `rails g active_admin:install` instead. I'll try to move forward with this again.

**Comment by vfonic** [comment]
I'd love to help. I'm not sure about the time available that I'll have, but let's see. Is there any repo where I can have a look? Maybe we could split some work?

**Comment by Fire-Dragon-DoL** [comment]
I have been going through the various issues to figure out if I could remove jquery as a dependency and realized how painful must have been doing the work to migrate to webpacker, just to see it deprecated.

I wonder if activeadmin should do the following:
- offer an npm package with all the css, js and assets bundled, so that anybody can integrate that in their JS pipeline. Ideally provide these on a CDN too (importmap and all that) AND
- use whatever is the default rails pipeline to provide css and js without the npm package, but as a separate gem (essentially allowing an opt in)

This way it should be possible to make everybody happy:
- drop activeadmin into gemfile + activeadmin-js (just came up with it), everything works
- drop activeadmin in gemfile and manage JS however you want with the npm package. This in turn will provide "natural support" to importmap

The downside is that this will effectively need to be a breaking release, because to keep activeadmin working, people will need to add a gem to their gemfile

As for dropping jquery and jquery ui, I'll leave that open. The reality that a good chunk of the js community moved on it's there, which makes it more challenging finding premade components without importing another library (incidentally, react)

A solution to this would be to use custom html elements as an interface and provide an npm package that's essentially just implementing this html elements using the js code.
Now, this html elements definition is public API (versioning and all that), but people can go off and replace the entirety of activeadmin JS if they want.

Would love to hear thoughts on this

**Comment by javierjulio** [comment]
This has already been removed on the default branch for the next major release so closing this out.

We've addressed the issue around date pickers by using the native date picker instead which also works great on mobile. Users can target that element using JS if they rather change it to a custom date picker. The default though will be the native date picker.

We've removed the batch action form DSL and has been replaced with a partial/modal-id feature. This will give full control and arguably a better DSL for batch action modals since we'd be able to use Rails forms. Modal functionality now comes from Flowbite.

**Reply by Fire-Dragon-DoL** [reply]
Wow! amazing!
